### PR TITLE
DAOS-6099 dfuse: Add fstat to interception library.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a wrapper around fstat calls so they can be intercepted and use os-bypass
to talk direct to the servers themselves.
Update the ioctl to return the caching state of the file and only use this new
feature if metadata caching is disabled, if it's enabled and there are multiple
processes per node or frequent operations it's likely quicker to use the kernel
cache.

### CODE REVIEW

**Review by mchaarawi — DISMISSED** [review]
PR does have unrelated things in it, but looks fine to me.

File: src/client/dfuse/il/int_posix.c [hunk 1]
```diff
@@ -1404,6 +1409,62 @@ dfuse_fclose(FILE *stream)
...
+dfuse___fxstat(int ver, int fd, struct stat *buf)
+{
+	struct fd_entry *entry = NULL;
+	int rc;
+
+	rc = vector_get(&fd_table, fd, &entry);
+	if (rc != 0)
+		goto do_real_fstat;
+
+	/* Turn of this feature if the kernel is doing metadata caching, in this case it's btter
```
Comment (daltonbohning) [inline src/client/dfuse/il/int_posix.c:1422]: Only if you need to push again
```suggestion
	/* Turn off this feature if the kernel is doing metadata caching, in this case it's btter
```

File: src/client/dfuse/il/int_posix.c [hunk 2]
```diff
@@ -1404,6 +1409,62 @@ dfuse_fclose(FILE *stream)
...
+		}
+		entry->fd_dev = buf->st_dev;
+		entry->fd_ino = buf->st_ino;
+		vector_decref(&fd_table, entry);
+		return 0;
+	}
+
+	rc = dfs_ostat(entry->fd_cont->ioc_dfs, entry->fd_dfsoh, buf);
+
+	DFUSE_TRA_INFO(entry->fd_dfsoh, "fstat() returned %d", rc);
```
Comment (daltonbohning) [inline src/client/dfuse/il/int_posix.c:1451]: Minor: would this make more sense?
```suggestion
	rc = dfs_ostat(entry->fd_cont->ioc_dfs, entry->fd_dfsoh, buf);

	DFUSE_TRA_INFO(entry->fd_dfsoh, "dfs_ostat() returned %d", rc);
```

**Review by daltonbohning — DISMISSED** [review]

File: src/client/dfuse/il/int_posix.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/il/int_posix.c:1422]: I got so confused with caching vs not caching and when the feature was on or off when writing this PR, maybe this is why.  I think we probably *don't* want this in TB3 so I'll likely need to re-push and will fix this then.

**Review by jolivier23 — DISMISSED** [review]

**Comment by jolivier23** [comment]
I'll defer to you to land.  I couldn't figure out what to put for commit message.  Also, you indicated that you might repush anyway

**Comment by ashleypittman** [comment]
> I'll defer to you to land. I couldn't figure out what to put for commit message. Also, you indicated that you might repush anyway

I landed a different IL patch so had to resolve conflicts, and I have added the notification checks to fstat as well.

**Review by jolivier23 — APPROVED** [review]

**Review by daltonbohning — APPROVED** [review]

**Comment by ashleypittman** [comment]
@daos-stack/daos-gatekeeper I've updated the PR description with a commit message.

**Pull request merged**